src/main/kotlin/teaforge/platform/RoboRio/Main.kt

```diff
@@ -1,20 +1,68 @@
 package teaforge.platform.RoboRio
 
+import com.ctre.phoenix6.hardware.CANcoder
+import com.ctre.phoenix6.hardware.Pigeon2
+import com.ctre.phoenix6.hardware.TalonFX
+import com.revrobotics.spark.SparkMax
+import edu.wpi.first.math.geometry.Rotation3d
 import edu.wpi.first.wpilibj.RobotBase
+import teaforge.ProgramConfig
 import teaforge.platform.RoboRio.internal.TimedRobotBasedPlatform
+import teaforge.utils.Maybe
+import teaforge.utils.Result
+import java.util.*
 
 fun <TMessage, TModel> timedRobotProgram(program: RoboRioProgram<TMessage, TModel>): RobotBase {
     return TimedRobotBasedPlatform<TMessage, TModel>(program)
 }
 
 typealias RoboRioProgram<TMessage, TModel> =
-        teaforge.ProgramConfig<Effect<TMessage>, TMessage, TModel, Subscription<TMessage>>
+        ProgramConfig<Effect<TMessage>, TMessage, TModel, Subscription<TMessage>>
 
 sealed interface Effect<out TMessage> {
     data class Log(val msg: String) : Effect<Nothing>
 
+    data class LoadSong<TMessage>(
+        val motor: CanDeviceToken.MotorToken.TalonMotorToken,
+        val songData: ByteArray,
+        val message: (CanDeviceToken.MotorToken.TalonMotorToken, Maybe<Error>) -> TMessage
+    ) : Effect<TMessage>
+
+    data class PlaySong<TMessage>(
+        val motor: CanDeviceToken.MotorToken.TalonMotorToken,
+        val message: (CanDeviceToken.MotorToken.TalonMotorToken, Maybe<Error>) -> TMessage
+    ) : Effect<TMessage>
+
+    data class StopSong<TMessage>(
+        val motor: CanDeviceToken.MotorToken.TalonMotorToken,
+        val message: (CanDeviceToken.MotorToken.TalonMotorToken, Maybe<Error>) -> TMessage
+    ) : Effect<TMessage>
+
     data class SetPwmMotorSpeed<TMessage>(val pwmSlot: PwmPort, val value: Double) :
-            Effect<TMessage>
+        Effect<TMessage>
+
+    data class InitCanDevice<TMessage>(
+        val type: CanDeviceType,
+        val id: Int,
+        val message: (InitCanDevice<TMessage>, Result<CanDeviceToken<*>, Error>) -> TMessage
```

Comment:
Instead of returning the original message we could just send the CanDeviceType and id.